jena-arq/src/main/java/org/apache/jena/sparql/algebra/optimize/TransformFilterImplicitJoin.java

```diff
@@ -120,10 +120,14 @@ private static Op apply(ExprList exprs, Op subOp) {
         // ---- Check if the subOp is the right shape to transform.
         Op op = subOp;
 
-        // Special case : deduce that the filter will always "eval unbound"
-        // hence eliminate all rows. Return the empty table.
+        // LATERAL : This is not longer true.
+//        // Special case : deduce that the filter will always "eval unbound"
+//        // hence eliminate all rows. Return the empty table.
+//        if (testSpecialCaseUnused(subOp, joins, remaining))
+//            return OpTable.empty();
+        // But simply skipping this causes (filter) to become (assign) which fails as (assign) does not handle errors.
         if (testSpecialCaseUnused(subOp, joins, remaining))
-            return OpTable.empty();
+            return null;
```

Comment:
Would an alternative fix be overriding the transform of `OpLateral` in this class to return a direct copy of the original `opLateral` ignoring any transforms this might have generated?

Reply:
That's a possibility - I think it seems to designing for handling a bad query case (if not LATERAL), rather than optimization in LATERAL.

It does seem to be a risk more widely, now and in future changes and not just in LATERAL - e.g. optimizing then substituting (e.g. parameterized queries). (NOT)EXISTS as well which is related to substitution. The fix for JENA-500 is a specific fix for the whole-query case.

The transform walk is bottom up so it will run over the code - the walker needs to be changed to avoid certain ops.

Reply:
Yes the bottom up property means the optimisation would be attempted but then you could discard it when you reached the `OpLateral` level but that does mean unnecessary work is done.

Having the walker be able to decide what operations to descend through could be a useful mechanism in general for limiting the scope of transformations

Reply:
Changing the walker is big change - I've tried it before. I'd like to get this PR in without resorting to that.

I believe the special case is also a potential impact in other places. We haven't seen it reported because e.g. EXISTS {} is often simple.

Is there any pointers as to when it has been beneficial?

I tried skipping it rather than `return null` but `TestSemanticEquivalence.implicitJoinEvaluation3` and on a SPARQL WG test fail.

In `implicitJoinEvaluation3`, it only fails on the query string part - the next two algebra level tests work. This doesn't make sense to me -- the algebra steps are supposed to cover the possibilities of the SPARQL query.

Reply:
Query timeouts will catch that (we hope).

EXISTS uses the same injection of bindings as JENA-500 except it does it dynamically at runtime (like LATERAL - EXIST is the ASK version of LATERAL). Whether that's the current iffy form or the better one written up.

It's is less likely to occur in EXISTS due to usage but it is possible. And like LATERAL, complicated because it may be bound sometimes and not others.

Can we merge as-is which is functionally safe?

---

I'm still mystified by `TestSemanticEquivalence.implicitJoinEvaluation3`.

Maybe the test is wrong and comparing to zero happens for the wrong reasons but I can't see a difference to the algebra sub-tests.